README.md
# Documentation for PyQuadTree

A pure Python QuadTree spatial index for GIS or rendering usage.

Karim Bahgat, 2014

Version: 0.2

License: MIT

Platforms: Python 2x and 3x

Based on Matt Rasmussen's original code:
https://github.com/mdrasmus/compbio/blob/master/rasmus/quadtree.py

## Installing It

Installing PyQuadTree is as easy as putting the "pyqtree" folder anywhere Python can import it,
such as the folder "PythonXX/Lib/site-packages"

## Example Usage

Start your session by importing the module.

```python
import pyqtree
```

Setup the spatial index, giving it a bounding box area to keep track of.
The bounding box being in a four-tuple: (xmin,ymin,xmax,ymax).

```python
spindex = pyqtree.Index(bbox=[0,0,100,100])
```

Populate the index with items that you want to be retrieved at a later point,
along with each item's geographic bbox.

```python
#this example assumes you have a list of items with bbox attribute
for item in items:
    spindex.insert(item=item, bbox=item.bbox)
```

Then when you have a region of interest and you wish to retrieve items from that region,
just use the index's intersect method. This quickly gives you a list of the stored items
whose bboxes intersects your region of interests. 

```python
overlapbbox = (51,51,86,86)
matches = spindex.intersect(overlapbbox)
```

There are other things that can be done as well, but that's it for the main usage!


## Functions and Classes

### pyqtree.Index(...) --> class object
The top spatial index to be created by the user. Once created it can be
populated with geographically placed members that can later be tested for
intersection with a user inputted geographic bounding box. Note that the
index can be iterated through in a for-statement, which loops through all
all the quad instances and lets you access their properties.

| **option** | **description**
| --- | --- 
| bbox | the coordinate system bounding box of the area that the quadtree should keep track of, as a 4-length sequence (xmin,ymin,xmax,ymax)

  - #### .countmembers(...):
  Returns a count of the total number of members/items/nodes inserted into
  this quadtree and all of its child trees.

  - #### .insert(...):
  Inserts an item into the quadtree along with its bounding box.
  
  | **option** | **description**
  | --- | --- 
  | item | the item to insert into the index, which will be returned by the intersection method
  | bbox | the spatial bounding box tuple of the item, with four members (xmin,ymin,xmax,ymax)

  - #### .intersect(...):
  Intersects an input boundingbox rectangle with all of the items
  contained in the quadtree. Returns a list of items whose bounding
  boxes intersect with the input rectangle.
  
  | **option** | **description**
  | --- | --- 
  | bbox | a spatial bounding box tuple with four members (xmin,ymin,xmax,ymax)
  | results | only used internally



CHANGES.txt

## Changes

### 1.0.0 (2018-09-14)

- Bump to first major version
- Fix so returns list instead of set
- Support inserting hashable items

### 0.25.0 (2016-06-22)

- Misc user contributions and bug fixes

### 0.24.0 (2015-06-18)

- Previous stable PyPI version.


pyqtree_test.py
from pyqtree import Index

INDEX_BBOX = (0, 0, 10, 10)

ITEM1 = 'Item 1'
BBOX1 = (0, 0, 0, 0)

ITEM2 = {'name': 'Item 2'} # non-hashable
BBOX2 = (1, 1, 1, 1)


def test_should_initially_count_zero_members():
    index = Index(INDEX_BBOX)
    assert len(index) == 0

def test_should_add_single_node_and_find_its_intersection():
    index = Index(INDEX_BBOX)
    index.insert(ITEM1, BBOX1)
    assert len(index) == 1
    assert index.intersect(BBOX1) == [ITEM1]

def test_should_add_multiple_nodes_and_find_them():
    index = Index(INDEX_BBOX)
    index.insert(ITEM1, BBOX1)
    index.insert(ITEM2, BBOX2)
    assert len(index) == 2
    assert index.intersect(BBOX1) == [ITEM1]
    assert index.intersect(BBOX2) == [ITEM2]
    res = index.intersect(INDEX_BBOX)
    assert ITEM1 in res
    assert ITEM2 in res

def test_should_empty_index_after_removing_added_node():
    index = Index(INDEX_BBOX)
    index.insert(ITEM1, BBOX1)
    index.remove(ITEM1, BBOX1)
    assert len(index) == 0
    assert index.intersect(BBOX1) == []

def test_should_empty_index_after_removing_multiple_added_nodes_in_separate_quad_nodes():
    index = Index(INDEX_BBOX, max_items=1)
    index.insert(ITEM1, BBOX1)
    index.insert(ITEM2, BBOX2)
    index.remove(ITEM1, BBOX1)
    index.remove(ITEM2, BBOX2)
    assert len(index) == 0
    assert index.intersect(INDEX_BBOX) == []

def test_should_empty_index_after_removing_multiple_added_nodes_in_all_quad_nodes():
    index = Index(INDEX_BBOX, max_items=1)
    index.insert(ITEM1, BBOX1)
    index.insert(ITEM2, INDEX_BBOX)
    index.remove(ITEM1, BBOX1)
    index.remove(ITEM2, INDEX_BBOX)
    assert len(index) == 0
    assert index.intersect(INDEX_BBOX) == []

def test_should_empty_index_after_removing_multiple_added_nodes_in_multiple_horizontal_quad_nodes():
    index = Index(INDEX_BBOX, max_items=1)
    bbox2 = (0, 0, INDEX_BBOX[2], 1)
    index.insert(ITEM1, BBOX1)
    index.insert(ITEM2, bbox2)
    index.remove(ITEM1, BBOX1)
    index.remove(ITEM2, bbox2)
    assert len(index) == 0
    assert index.intersect(INDEX_BBOX) == []

def test_should_empty_index_after_removing_multiple_added_nodes_in_multiple_vertical_quad_nodes():
    index = Index(INDEX_BBOX, max_items=1)
    bbox2 = (0, 0, 1, INDEX_BBOX[3])
    index.insert(ITEM1, BBOX1)
    index.insert(ITEM2, bbox2)
    index.remove(ITEM1, BBOX1)
    index.remove(ITEM2, bbox2)
    assert len(index) == 0
    assert index.intersect(INDEX_BBOX) == []

if __name__ == '__main__':
    import traceback
    tests = [(name,obj) for name,obj in list(globals().items()) if hasattr(obj, '__call__') and obj.__name__.startswith('test_')]
    failed = 0
    for name,obj in tests:
        try:
            obj()
        except Exception as err:
            failed += 1
            print('test fail: "{}"'.format(name))
            traceback.print_exc()

    if failed:
        print('{} out of {} tests failed'.format(failed, len(tests)))
    else:
        print('All tests completed successfully')


upload.py
import pipy
 
packpath = "pyqtree.py"
pipy.define_upload(packpath,
                   author="Karim Bahgat",
                   author_email="karim.bahgat.norway@gmail.com",
                   license="MIT",
                   name="Pyqtree",
                   description="A pure Python quad tree spatial index for GIS or rendering usage.",
                   url="http://github.com/karimbahgat/Pyqtree",
                   keywords="GIS spatial index quad tree",
                   classifiers=["License :: OSI Approved",
                                "Programming Language :: Python",
                                "Development Status :: 4 - Beta",
                                "Intended Audience :: Developers",
                                "Intended Audience :: Science/Research",
                                'Intended Audience :: End Users/Desktop',
                                "Topic :: Scientific/Engineering :: GIS"],
                   changes=["Bump to first major version",
                            "Fix so returns list instead of set",
                            "Support inserting hashable items",
                            ],
                   )

pipy.generate_docs(packpath)
#pipy.upload_test(packpath)
#pipy.upload(packpath)



pyqtree.py
"""
# Pyqtree

Pyqtree is a pure Python spatial index for GIS or rendering usage.
It stores and quickly retrieves items from a 2x2 rectangular grid area,
and grows in depth and detail as more items are added.
The actual quad tree implementation is adapted from
[Matt Rasmussen's compbio library](https://github.com/mdrasmus/compbio/blob/master/rasmus/quadtree.py)
and extended for geospatial use.


## Platforms

Python 2 and 3.


## Dependencies

Pyqtree is written in pure Python and has no dependencies.


## Installing It

Installing Pyqtree can be done by opening your terminal or commandline and typing:

    pip install pyqtree

Alternatively, you can simply download the "pyqtree.py" file and place
it anywhere Python can import it, such as the Python site-packages folder.


## Example Usage

Start your script by importing the quad tree.

    from pyqtree import Index

Setup the spatial index, giving it a bounding box area to keep track of.
The bounding box being in a four-tuple: (xmin, ymin, xmax, ymax).

    spindex = Index(bbox=(0, 0, 100, 100))

Populate the index with items that you want to be retrieved at a later point,
along with each item's geographic bbox.

    # this example assumes you have a list of items with bbox attribute
    for item in items:
        spindex.insert(item, item.bbox)

Then when you have a region of interest and you wish to retrieve items from that region,
just use the index's intersect method. This quickly gives you a list of the stored items
whose bboxes intersects your region of interests.

    overlapbbox = (51, 51, 86, 86)
    matches = spindex.intersect(overlapbbox)

There are other things that can be done as well, but that's it for the main usage!


## More Information:

- [Home Page](http://github.com/karimbahgat/Pyqtree)
- [API Documentation](https://karimbahgat.github.io/Pyqtree/)


## License:

This code is free to share, use, reuse, and modify according to the MIT license, see LICENSE.txt.


## Credits:

- Karim Bahgat
- Joschua Gandert

"""

__version__ = "1.0.0"

#PYTHON VERSION CHECK
import sys
PYTHON3 = int(sys.version[0]) == 3
if PYTHON3:
    xrange = range


def _normalize_rect(rect):
    if len(rect) == 2:
        x1, y1 = rect
        x2, y2 = rect
    else:
        x1, y1, x2, y2 = rect
    if x1 > x2:
        x1, x2 = x2, x1
    if y1 > y2:
        y1, y2 = y2, y1
    return (x1, y1, x2, y2)


def _loopallchildren(parent):
    for child in parent.children:
        if child.children:
            for subchild in _loopallchildren(child):
                yield subchild
        yield child


class _QuadNode(object):
    def __init__(self, item, rect):
        self.item = item
        self.rect = rect

    def __eq__(self, other):
        return self.item == other.item and self.rect == other.rect

    def __hash__(self):
        return hash(self.item)


class _QuadTree(object):
    """
    Internal backend version of the index.
    The index being used behind the scenes. Has all the same methods as the user
    index, but requires more technical arguments when initiating it than the
    user-friendly version.
    """

    def __init__(self, x, y, width, height, max_items, max_depth, _depth=0):
        self.nodes = []
        self.children = []
        self.center = (x, y)
        self.width, self.height = width, height
        self.max_items = max_items
        self.max_depth = max_depth
        self._depth = _depth

    def __iter__(self):
        for child in _loopallchildren(self):
            yield child

    def _insert(self, item, bbox):
        rect = _normalize_rect(bbox)
        if len(self.children) == 0:
            node = _QuadNode(item, rect)
            self.nodes.append(node)

            if len(self.nodes) > self.max_items and self._depth < self.max_depth:
                self._split()
        else:
            self._insert_into_children(item, rect)

    def _remove(self, item, bbox):
        rect = _normalize_rect(bbox)
        if len(self.children) == 0:
            node = _QuadNode(item, rect)
            self.nodes.remove(node)
        else:
            self._remove_from_children(item, rect)

    def _intersect(self, rect, results=None, uniq=None):
        if results is None:
            rect = _normalize_rect(rect)
            results = []
            uniq = set()
        # search children
        if self.children:
            if rect[0] <= self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[0]._intersect(rect, results, uniq)
                if rect[3] >= self.center[1]:
                    self.children[1]._intersect(rect, results, uniq)
            if rect[2] >= self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[2]._intersect(rect, results, uniq)
                if rect[3] >= self.center[1]:
                    self.children[3]._intersect(rect, results, uniq)
        # search node at this level
        for node in self.nodes:
            _id = id(node.item)
            if (_id not in uniq and
                node.rect[2] >= rect[0] and node.rect[0] <= rect[2] and
                node.rect[3] >= rect[1] and node.rect[1] <= rect[3]):
                results.append(node.item)
                uniq.add(_id)
        return results

    def _insert_into_children(self, item, rect):
        # if rect spans center then insert here
        if (rect[0] <= self.center[0] and rect[2] >= self.center[0] and
            rect[1] <= self.center[1] and rect[3] >= self.center[1]):
            node = _QuadNode(item, rect)
            self.nodes.append(node)
        else:
            # try to insert into children
            if rect[0] <= self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[0]._insert(item, rect)
                if rect[3] >= self.center[1]:
                    self.children[1]._insert(item, rect)
            if rect[2] > self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[2]._insert(item, rect)
                if rect[3] >= self.center[1]:
                    self.children[3]._insert(item, rect)

    def _remove_from_children(self, item, rect):
        # if rect spans center then insert here
        if (rect[0] <= self.center[0] and rect[2] >= self.center[0] and
            rect[1] <= self.center[1] and rect[3] >= self.center[1]):
            node = _QuadNode(item, rect)
            self.nodes.remove(node)
        else:
            # try to remove from children
            if rect[0] <= self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[0]._remove(item, rect)
                if rect[3] >= self.center[1]:
                    self.children[1]._remove(item, rect)
            if rect[2] > self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[2]._remove(item, rect)
                if rect[3] >= self.center[1]:
                    self.children[3]._remove(item, rect)

    def _split(self):
        quartwidth = self.width / 4.0
        quartheight = self.height / 4.0
        halfwidth = self.width / 2.0
        halfheight = self.height / 2.0
        x1 = self.center[0] - quartwidth
        x2 = self.center[0] + quartwidth
        y1 = self.center[1] - quartheight
        y2 = self.center[1] + quartheight
        new_depth = self._depth + 1
        self.children = [_QuadTree(x1, y1, halfwidth, halfheight,
                                   self.max_items, self.max_depth, new_depth),
                         _QuadTree(x1, y2, halfwidth, halfheight,
                                   self.max_items, self.max_depth, new_depth),
                         _QuadTree(x2, y1, halfwidth, halfheight,
                                   self.max_items, self.max_depth, new_depth),
                         _QuadTree(x2, y2, halfwidth, halfheight,
                                   self.max_items, self.max_depth, new_depth)]
        nodes = self.nodes
        self.nodes = []
        for node in nodes:
            self._insert_into_children(node.item, node.rect)

    def __len__(self):
        """
        Returns:

        - A count of the total number of members/items/nodes inserted
        into this quadtree and all of its child trees.
        """
        size = 0
        for child in self.children:
            size += len(child)
        size += len(self.nodes)
        return size


MAX_ITEMS = 10
MAX_DEPTH = 20


class Index(_QuadTree):
    """
    The top spatial index to be created by the user. Once created it can be
    populated with geographically placed members that can later be tested for
    intersection with a user inputted geographic bounding box. Note that the
    index can be iterated through in a for-statement, which loops through all
    all the quad instances and lets you access their properties.

    Example usage:

    >>> spindex = Index(bbox=(0, 0, 100, 100))
    >>> spindex.insert('duck', (50, 30, 53, 60))
    >>> spindex.insert('cookie', (10, 20, 15, 25))
    >>> spindex.insert('python', (40, 50, 95, 90))
    >>> results = spindex.intersect((51, 51, 86, 86))
    >>> sorted(results)
    ['duck', 'python']
    """

    def __init__(self, bbox=None, x=None, y=None, width=None, height=None, max_items=MAX_ITEMS, max_depth=MAX_DEPTH):
        """
        Initiate by specifying either 1) a bbox to keep track of, or 2) with an xy centerpoint and a width and height.

        Parameters:
        - **bbox**: The coordinate system bounding box of the area that the quadtree should
            keep track of, as a 4-length sequence (xmin,ymin,xmax,ymax)
        - **x**:
            The x center coordinate of the area that the quadtree should keep track of.
        - **y**
            The y center coordinate of the area that the quadtree should keep track of.
        - **width**:
            How far from the xcenter that the quadtree should look when keeping track.
        - **height**:
            How far from the ycenter that the quadtree should look when keeping track
        - **max_items** (optional): The maximum number of items allowed per quad before splitting
            up into four new subquads. Default is 10.
        - **max_depth** (optional): The maximum levels of nested subquads, after which no more splitting
            occurs and the bottommost quad nodes may grow indefinately. Default is 20.
        """
        if bbox is not None:
            x1, y1, x2, y2 = bbox
            width, height = abs(x2-x1), abs(y2-y1)
            midx, midy = x1+width/2.0, y1+height/2.0
            super(Index, self).__init__(midx, midy, width, height, max_items, max_depth)

        elif None not in (x, y, width, height):
            super(Index, self).__init__(x, y, width, height, max_items, max_depth)

        else:
            raise Exception("Either the bbox argument must be set, or the x, y, width, and height arguments must be set")

    def insert(self, item, bbox):
        """
        Inserts an item into the quadtree along with its bounding box.

        Parameters:
        - **item**: The item to insert into the index, which will be returned by the intersection method
        - **bbox**: The spatial bounding box tuple of the item, with four members (xmin,ymin,xmax,ymax)
        """
        self._insert(item, bbox)

    def remove(self, item, bbox):
        """
        Removes an item from the quadtree.

        Parameters:
        - **item**: The item to remove from the index
        - **bbox**: The spatial bounding box tuple of the item, with four members (xmin,ymin,xmax,ymax)

        Both parameters need to exactly match the parameters provided to the insert method.
        """
        self._remove(item, bbox)

    def intersect(self, bbox):
        """
        Intersects an input boundingbox rectangle with all of the items
        contained in the quadtree.

        Parameters:
        - **bbox**: A spatial bounding box tuple with four members (xmin,ymin,xmax,ymax)

        Returns:
        - A list of inserted items whose bounding boxes intersect with the input bbox.
        """
        return self._intersect(bbox)


testing.py

import pyqtree
import random, time

class Item:
    def __init__(self, x, y):
        left = x-1
        right = x+1
        top = y-1
        bottom = y+1
        self.bbox = [left,top,right,bottom]

#setup and populate index
items = [Item(random.randrange(5,95),random.randrange(5,95)) for _ in range(10000)]
spindex = pyqtree.Index(bbox=[-11,-33,100,100])
for item in items:
    spindex.insert(item, item.bbox)
print("{0} members in this index.".format(spindex.countmembers()))

#test intersection
print("testing hit")
testitem = (51,51,86,86)
t = time.time()
matches = spindex.intersect(testitem)
print(time.time()-t, " seconds")

#test countmembers()
# trivial list of items
items = [Item(0.5, 0.5), Item(-0.5, 0.5), Item(-0.5, -0.5), Item(0.5, -0.5)]

# populate: maxindex=3, so we must split
spindex = pyqtree.Index(bbox=[-1, -1, 1, 1], maxitems=3)
for item in items:
    spindex.insert(item, item.bbox)

# check result
members = spindex.countmembers()
assert(members == 4)
print("{0} nodes in this Index.".format(members))



benchmarking.py

import pyqtree
import random, time

class Item:
    def __init__(self, x, y):
        left = x-1
        right = x+1
        top = y-1
        bottom = y+1
        self.bbox = [left,top,right,bottom]

#setup and populate index
items = [Item(random.randrange(5,95),random.randrange(5,95)) for _ in range(10000)]
spindex = pyqtree.Index(bbox=[-11,-33,100,100])
for item in items:
    spindex.insert(item, item.bbox)
print("{0} members in this index.".format(len(spindex)))

#test intersection
print("testing hit")
testitem = (51,51,86,86)
t = time.time()
matches = spindex.intersect(testitem)
print(len(matches))
print(len(set(matches)))
print("{0} seconds".format(time.time()-t))

#test countmembers()
# trivial list of items
items = [Item(0.5, 0.5), Item(-0.5, 0.5), Item(-0.5, -0.5), Item(0.5, -0.5)]

# populate: maxindex=3, so we must split
spindex = pyqtree.Index(bbox=[-1, -1, 1, 1], max_items=3)
for item in items:
    spindex.insert(item, item.bbox)

# check result
members = len(spindex)
assert(members == 4)
print("{0} nodes in this Index.".format(members))



pyqtree/__init__.py
"""
PyQuadTree

A pure Python QuadTree spatial index for GIS or rendering usage.

Karim Bahgat, 2014

Version: 0.2

License: MIT

Platforms: Python 2x and 3x

Based on Matt Rasmussen's original code:
https://github.com/mdrasmus/compbio/blob/master/rasmus/quadtree.py

## Installing It

Installing PyQuadTree is as easy as putting the "pyqtree" folder anywhere Python can import it,
such as the folder "PythonXX/Lib/site-packages"

## Example Usage

Start your session by importing the module.

```python
import pyqtree
```

Setup the spatial index, giving it a bounding box area to keep track of.
The bounding box being in a four-tuple: (xmin,ymin,xmax,ymax).

```python
spindex = pyqtree.Index(bbox=[0,0,100,100])
```

Populate the index with items that you want to be retrieved at a later point,
along with each item's geographic bbox.

```python
#this example assumes you have a list of items with bbox attribute
for item in items:
    spindex.insert(item=item, bbox=item.bbox)
```

Then when you have a region of interest and you wish to retrieve items from that region,
just use the index's intersect method. This quickly gives you a list of the stored items
whose bboxes intersects your region of interests. 

```python
overlapbbox = (51,51,86,86)
matches = spindex.intersect(overlapbbox)
```

There are other things that can be done as well, but that's it for the main usage!

"""

#PYTHON VERSION CHECK
import sys
PYTHON3 = int(sys.version[0]) == 3
if PYTHON3:
    xrange = range

#INTERNAL USE ONLY
def _normalize_rect(rect):
    x1, y1, x2, y2 = rect
    if x1 > x2:
        x1, x2 = x2, x1
    if y1 > y2:
        y1, y2 = y2, y1
    return (x1, y1, x2, y2)

class _QuadNode:    
    def __init__(self, item, rect):
        self.item = item
        self.rect = rect
        
class _Index:
    """
    The index being used behind the scenes. Has all the same methods as the user
    index, but requires more technical arguments when initiating it than the
    user-friendly version. 

    | **option** | **description**
    | --- | --- 
    | x | the x center coordinate of the area that the quadtree should keep track of
    | y | the y center coordinate of the area that the quadtree should keep track of
    | width | how far from the xcenter that the quadtree should look when keeping track
    | height | how far from the ycenter that the quadtree should look when keeping track

    """
    MAX = 10
    MAX_DEPTH = 20
    def __init__(self, x, y, width, height, depth = 0):
        self.nodes = []
        self.children = []
        self.center = [x, y]
        self.width,self.height = width,height
        self.depth = depth
    def __iter__(self):
        def loopallchildren(parent):
            for child in parent.children:
                if child.children:
                    for subchild in loopallchildren(parent=child):
                        yield subchild
                yield child
        for child in loopallchildren(self):
            yield child
    def insert(self, item, bbox):
        """
        Inserts an item into the quadtree along with its bounding box.

        | **option** | **description**
        | --- | --- 
        | item | the item to insert into the index, which will be returned by the intersection method
        | bbox | the spatial bounding box tuple of the item, with four members (xmin,ymin,xmax,ymax)
        """
        rect = _normalize_rect(bbox)
        if len(self.children) == 0:
            node = _QuadNode(item, rect)
            self.nodes.append(node)
            
            if len(self.nodes) > self.MAX and self.depth < self.MAX_DEPTH:
                self._split()
        else:
            self._insert_into_children(item, rect)
    def intersect(self, bbox, results=None):
        """
        Intersects an input boundingbox rectangle with all of the items
        contained in the quadtree. Returns a list of items whose bounding
        boxes intersect with the input rectangle.

        | **option** | **description**
        | --- | --- 
        | bbox | a spatial bounding box tuple with four members (xmin,ymin,xmax,ymax)
        | results | only used internally
        """
        rect = bbox
        if results is None:
            rect = _normalize_rect(rect)
            results = set()
        # search children
        if len(self.children) > 0:
            if rect[0] <= self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[0].intersect(rect, results)
                if rect[3] > self.center[1]:
                    self.children[1].intersect(rect, results)
            if rect[2] > self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[2].intersect(rect, results)
                if rect[3] > self.center[1]:
                    self.children[3].intersect(rect, results)
        # search node at this level
        for node in self.nodes:
            if (node.rect[2] > rect[0] and node.rect[0] <= rect[2] and 
                node.rect[3] > rect[1] and node.rect[1] <= rect[3]):
                results.add(node.item)
        return results
    def countmembers(self):
        """
        Returns a count of the total number of members/items/nodes inserted into
        this quadtree and all of its child trees.
        """
        size = 0
        for child in self.children:
            size += child.countmembers()
        size += len(self.nodes)
        return size
    
    #INTERNAL USE ONLY
    def _insert_into_children(self, item, rect):
        # if rect spans center then insert here
        if ((rect[0] <= self.center[0] and rect[2] > self.center[0]) and
            (rect[1] <= self.center[1] and rect[3] > self.center[1])):
            node = _QuadNode(item, rect)
            self.nodes.append(node)
        else:
            # try to insert into children
            if rect[0] <= self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[0].insert(item, rect)
                if rect[3] > self.center[1]:
                    self.children[1].insert(item, rect)
            if rect[2] > self.center[0]:
                if rect[1] <= self.center[1]:
                    self.children[2].insert(item, rect)
                if rect[3] > self.center[1]:
                    self.children[3].insert(item, rect)
    def _split(self):
        quartwidth = self.width/4.0
        quartheight = self.height/4.0
        halfwidth = self.width/2.0
        halfheight = self.height/2.0
        self.children = [_Index(self.center[0] - quartwidth,
                                  self.center[1] - quartheight,
                                  width=halfwidth, height=halfheight,
                                  depth=self.depth + 1),
                         _Index(self.center[0] - quartwidth,
                                  self.center[1] + quartheight,
                                  width=halfwidth, height=halfheight,
                                  depth=self.depth + 1),
                         _Index(self.center[0] + quartwidth,
                                  self.center[1] - quartheight,
                                  width=halfwidth, height=halfheight,
                                  depth=self.depth + 1),
                         _Index(self.center[0] + quartwidth,
                                  self.center[1] + quartheight,
                                  width=halfwidth, height=halfheight,
                                  depth=self.depth + 1)]
        nodes = self.nodes
        self.nodes = []
        for node in nodes:
            self._insert_into_children(node.item, node.rect)

#USER CLASSES AND FUNCTIONS 
class Index(_Index):
    """
    The top spatial index to be created by the user. Once created it can be
    populated with geographically placed members that can later be tested for
    intersection with a user inputted geographic bounding box. Note that the
    index can be iterated through in a for-statement, which loops through all
    all the quad instances and lets you access their properties.

    | **option** | **description**
    | --- | --- 
    | bbox | the coordinate system bounding box of the area that the quadtree should keep track of, as a 4-length sequence (xmin,ymin,xmax,ymax)
    
    """
    def __init__(self, bbox):
        x1,y1,x2,y2 = bbox
        width,height = x2-x1,y2-y1
        midx,midy = x1+width/2.0, y1+height/2.0
        self.nodes = []
        self.children = []
        self.center = [midx, midy]
        self.width,self.height = width,height
        self.depth = 0

#SOME TESTING
if __name__ == "__main__":
    import random, time
    
    class Item:
        def __init__(self, x, y):
            left = x-1
            right = x+1
            top = y-1
            bottom = y+1
            self.bbox = [left,top,right,bottom]

    #setup and populate index
    items = [Item(random.randrange(5,95),random.randrange(5,95)) for _ in xrange(1000)]
    spindex = Index(bbox=[-11,-33,100,100])
    for item in items:
        spindex.insert(item, item.bbox)

    #test intersection
    print("testing hit")
    testitem = (51,51,86,86)
    t = time.time()
    matches = spindex.intersect(testitem)
    print(time.time()-t, " seconds")

